client/asset/dcr/dcr.go

```diff
@@ -1882,67 +1885,98 @@ func (dcr *ExchangeWallet) findRedemptionsInMempool(contractOutpoints []outPoint
 }
 
 // findRedemptionsInBlockRange attempts to find spending info for the specified
-// contracts by searching every input of all txs in the provided block range.
+// contracts by checking the v2cfilters of each block in the provided range for
+// likely inclusion of ANY of the specified contracts' P2SH script. If a block's
+// v2cfilters reports possible inclusion of ANY of the contracts' P2SH script,
+// all inputs of the matching block's txs are checked to determine if any of the
+// inputs spends any of the provided contracts.
 // If spending info is found for any contract, the contract is purged from the
 // findRedemptionQueue and the contract's secret (if successfully parsed) or any
 // error that occurs during parsing is returned to the redemption finder via the
 // registered result chan.
-// Also checks mempool for potential redemptions if spending info is not found
-// for any of these contracts in the specified block range.
-// TODO: Improve redemption search in mined blocks by scanning block filters
-// rather than every input of every tx in a block.
-func (dcr *ExchangeWallet) findRedemptionsInBlockRange(startBlock, endBlock *block, contractOutpoints []outPoint) {
-	contractsCount := len(contractOutpoints)
+// If spending info is not found for any of these contracts after checking the
+// specified block range, a mempool search is triggered to attempt finding unmined
+// redemptions for the remaining contracts.
+// NOTE:
+// Any error encountered while checking a block's cfilters or fetching a matching
+// block's txs compromises the redemption search for this set of contracts because
+// subsequent attempts to find these contracts' redemption will not repeat any
+// block in the specified range unless the contracts are first removed from the
+// findRedemptionQueue. Thus, any such error will cause this set of contracts to
+// be purged from the findRedemptionQueue. The error will be propagated to the
+// redemption finder(s) and these may re-call dcr.FindRedemption to restart find
+// redemption attempts for any of these contracts.
+func (dcr *ExchangeWallet) findRedemptionsInBlockRange(startBlockHeight, endBlockHeight int64, contractOutpoints []outPoint) {
+	totalContracts := len(contractOutpoints)
 	dcr.log.Debugf("finding redemptions for %d contracts in blocks %d - %d",
-		contractsCount, startBlock.height, endBlock.height)
+		totalContracts, startBlockHeight, endBlockHeight)
 
-	nextBlockHash := startBlock.hash
-	var lastScannedBlockHeight int64
+	var lastScannedBlockHeight int64 = -1 // we've not scanned any block yet
```

Comment:
`lastScannedBlockHeight` is only used for the `dcr.log.Debugf` after the loop, so there's little benefit from using a non-zero/default init afaict.  Having it -1 suggest to me I'm missing something or there's a logic error that would lead to the -1 value playing a role, but it just seems to be overwritten except on an early error.